### PR TITLE
[SPD-23149] Add filter that sums the elements of an array, optionally summing the values of a specified key in each object.

### PR DESCRIPTION
**Pull request by parth-sd**
## Description

Added `sumArray` filter: It either take an `array` of `objects` along with a `key` / an `array` of primitive datatypes & return the sum. 
The sum happens using the existing `performOperation` method (which is also used to do `plus`). So, it can handle boxed types (duration/date/currency) as well.
The `sumArray` will also handle errors like not an array or invalid key.

## Related Issue

https://spotdraft.atlassian.net/browse/SPD-23149

## Motivation and Context

This filter will be used to get the sum of all Column values of a Dynamic Table.

## How Has This Been Tested?

Unit Tests

### CODE REVIEW

File: filters.js [hunk 1]
```diff
@@ -55,6 +55,7 @@ var filters = {
   last: v => v[v.length - 1],
   lstrip: v => stringify(v).replace(/^\s+/, ""),
   map: (arr, arg) => arr.map(v => v[arg]),
+  sum: (arr, arg) => arr.reduce((acc, v) => acc + (v[arg] ?? v), 0),
```
Comment (sanket0896) [inline filters.js:58]: - What if the type is an object without having `arg` property?
- What if the value of the property is `undefined` or `null` or `NaN`?
- Add tests

Comment (codebreach) [reply filters.js:58]: +1 

also you need to handle boxed types (duration/date/currency)

Comment (parth-sd) [reply filters.js:58]: How should we handle for duration/currency? should we return the sum of their values only? @codebreach

File: test/filters.js [hunk 2]
```diff
@@ -46,7 +46,39 @@ var ctx = {
...
+  arr_invalid: null,
+  arr_empty: [],
+  arr_numbers: [1, 2, 3, 4, 5],
+  arr_numbers_with_mixed_datatypes: [undefined, 1, null, 2, {}, 3, 'a', 5, true],
+  dynamic_table_with_numbers: [{cost: 100, quantity: 2}, {cost: 200, quantity: 3}, {cost: 300, quantity: 4}],
+  dynamic_table_with_null: [{cost: 100, quantity: 2}, {cost: 200, quantity: 3}, {cost: null, quantity: 4}],
+  dynamic_table_with_empty: [{cost: 100, quantity: 2}, {cost: 200, quantity: 3}, {}]
+}
+
+ctx = {
```
Comment (sanket0896) [inline test/filters.js:59]: Add a comments saying why this is assigned separately

File: filters.js [hunk 3]
```diff
@@ -527,5 +527,41 @@ function toDuration(durValue, durType) {
...
+    } else if (_.isString(key)) {
+      // If a key is provided, map the values of that key from the array objects
+      const values = arr.map(item => item[key])
+
+      // Check if the key exists in all objects of the array
+      if (!values.every(isDefinedAndNotNullArg)) {
+        throw new Error("Key doesn't exist in all objects of array")
+      }
+
+      return values.slice(1).reduce((acc, item) => performOperations(acc, item, 'ADD'), values[0])
```
[the remark this answers is not in the document]
Comment (parth-sd) [reply filters.js:556]: oh, didn't knew that. thanks!

File: filters.js [hunk 4]
```diff
@@ -527,5 +527,41 @@ function toDuration(durValue, durType) {
...
+  } else if (arr.length > 0) {
+    if (key === undefined) {
+      // If no key is provided, sum the elements directly
+      return arr.slice(1).reduce((acc, item) => performOperations(acc, item, 'ADD'), arr[0])
+    } else if (_.isString(key)) {
+      // If a key is provided, map the values of that key from the array objects
+      const values = arr.map(item => item[key])
+
+      // Check if the key exists in all objects of the array
+      if (!values.every(isDefinedAndNotNullArg)) {
```
[the remark this answers is not in the document]
Comment (parth-sd) [reply filters.js:552]: Ok

**Review by sanket0896 — APPROVED** [review]

**Pull request merged**